### PR TITLE
#166214115 A registered user can reset his forgotten password

### PR DESCRIPTION
**Pull request by Adekoreday**
#### What does this PR do?
Enable a  registered user to reset his forgotten password via an email notification
#### Description of Task to be completed?
_add swagger documentation
#### How should this be manually tested?
```
clone the repository
npm install
add a .env file with the appropiate content as specified in the .env_sample file
npm run db:rollback
sign up as specified in the swagger docs
goto route localhost:3000/api/v1/auth/passwordreset and insert your registered mail in the body of the request
An email containing link to reset your password will be delivered to the registered mail 
click on the link to reset your password
```
#### What are the relevant pivotal tracker stories?
#166214115
#### Screenshots (if appropriate)
image

### CODE REVIEW

**Comment by allebd** [comment]
Kindly rebase

**Review by allebd — COMMENTED** [review]
Kindly rebase
having blocker expecting @Lundii to merge first
done

File: package.json [hunk 1]
```diff
@@ -9,7 +9,7 @@
     "start:dev": "nodemon --exec babel-node server/index.js",
     "heroku-postbuild": "npm run db:rollback && npm run build",
     "debug": "nodemon --exec babel-node server/index.js --inspect",
-    "test": "npm run db:roll && npm run db:seed && nyc mocha --recursive test --timeout 10000 -r esm --exit && npm run db:back",
+    "test": "npm run db:roll && npm run db:seed && nyc mocha --recursive test --timeout 20000 -r esm --exit && npm run db:back",
```
Comment (henryade) [inline package.json:12]: why did you change this to 20000

File: server/controllers/AuthController.js [hunk 2]
```diff
@@ -112,4 +111,56 @@ export default class AuthController {
...
   }
+
+/**
+ *
+ *
+ * @static
+ * @param {*} request
+ * @param {*} response
+ * @returns {object} - response object
+ * @memberof AuthController
```
Comment (henryade) [inline server/controllers/AuthController.js:119]: params should be object and returns json

File: server/controllers/AuthController.js [hunk 3]
```diff
@@ -112,4 +111,56 @@ export default class AuthController {
...
+ */
+  static async resetPassword(request, response) {
+
+    try {
+      const { id } = request.userData; parseInt(id, 10);
+      const user = await User.findOne({ where: { id } });
```
Comment (henryade) [inline server/controllers/AuthController.js]: use the helper function

File: server/controllers/AuthController.js [hunk 4]
```diff
@@ -112,4 +111,56 @@ export default class AuthController {
...
+ * @static
+ * @param {*} request
+ * @param {*} response
+ * @returns {object} - response object
+ * @memberof AuthController
+ */
+  static async verifyUserbyEmail(request, response) {
+    const { email } = request.body;
+    try {
+      const user = await User.findOne({ where: { email } });
```
Comment (henryade) [inline server/controllers/AuthController.js]: use the helper function

File: server/controllers/AuthController.js [hunk 5]
```diff
@@ -112,4 +111,56 @@ export default class AuthController {
...
+      if (user) {
+        const { id, dataValues } = user;
+        const { firstName } = dataValues;
+        const token = createToken({ id }, '2h');
+        const message = resetpasswordMessage(firstName, token);
+        await sendMail(process.env.ADMIN_MAIL, email, message);
+        return responseMessage(response, 200, { status: 'success', message: 'you will recieve a link in your mail shortly to proceed'});
+      }
+    } catch (error) {
+      return responseMessage(response, 500, { message: error.message });
```
Comment (henryade) [inline server/controllers/AuthController.js:134]: ignore this line from test

File: server/controllers/AuthController.js [hunk 6]
```diff
@@ -112,4 +111,56 @@ export default class AuthController {
...
+ * @param {*} response
+ * @returns {object} - response object
+ * @memberof AuthController
+ */
+  static async verifyUserbyEmail(request, response) {
+    const { email } = request.body;
+    try {
+      const user = await User.findOne({ where: { email } });
+      if (!user) return responseMessage(response, 404, { status: 'failure', message: 'user does not exist' });
+      if (user) {
```
Comment (henryade) [inline server/controllers/AuthController.js:125]: you dont need if statement here because it will only come here if there is a user

File: server/controllers/AuthController.js [hunk 7]
```diff
@@ -112,4 +111,56 @@ export default class AuthController {
...
+        const { id, dataValues } = user;
+        const { firstName } = dataValues;
```
Comment (henryade) [inline server/controllers/AuthController.js]: you can get firstname from the user also

**Review by henryade — CHANGES_REQUESTED** [review]

File: package.json (hunk 1 above)
Comment (Adekoreday) [reply package.json:12]: the test was failing due to timeouts, had to extend timeouts

File: server/controllers/AuthController.js (hunk 2 above)
Comment (Adekoreday) [reply server/controllers/AuthController.js:119]: noted

File: server/controllers/AuthController.js (hunk 4 above)
Comment (Adekoreday) [reply server/controllers/AuthController.js]: this seems a simpler its a single line, using the finduser helper will be joining tables and be providing data not needed

File: server/controllers/AuthController.js (hunk 6 above)
Comment (Adekoreday) [reply server/controllers/AuthController.js:125]: noted

File: server/controllers/AuthController.js (hunk 7 above)
Comment (Adekoreday) [reply server/controllers/AuthController.js]: noted thanks

File: server/controllers/AuthController.js [hunk 8]
```diff
@@ -112,4 +111,53 @@ export default class AuthController {
...
+      return responseMessage(response, 500, { message: error.message });
+    }        
+  }
+
+/**
+ *
+ *
+ * @static
+ * @param {*} request
+ * @param {*} response
```
Comment (Ukhu) [inline server/controllers/AuthController.js]: param should be object

File: server/routes/auth.js [hunk 9]
```diff
@@ -3,14 +3,17 @@ import AuthController from '../controllers/AuthController';
 
 import middlewares from '../middlewares';
 
-const { verifyToken, UserValidator } = middlewares;
```
Comment (Ukhu) [inline server/routes/auth.js]: no need to change this line. it's okay like this

File: server/controllers/AuthController.js [hunk 10]
```diff
@@ -112,4 +111,55 @@ export default class AuthController {
...
+/**
+ *
+ *
+ * @static
+ * @param {object} request
+ * @param {object} response
+ * @returns {json} - json
+ * @memberof AuthController
+ */
+  static async ResetpasswordEmail(request, response) {
```
Comment (henryade) [inline server/controllers/AuthController.js]: password should be capitalized also

File: server/controllers/AuthController.js [hunk 11]
```diff
@@ -56,7 +57,7 @@ export default class AuthController {
   /**
    * Method for handling signin route(POST api/v1/auth/signin)
    * @param {object} request - the request object
-   * @param {object} response  - the response object
+   * @param {json} response  - json
```
Comment (henryade) [inline server/controllers/AuthController.js]: response is object not json

File: server/controllers/AuthController.js [hunk 12]
```diff
@@ -106,4 +107,51 @@ export default class AuthController {
       return responseMessage(response, 500, { message: error.message });
```
Comment (henryade) [inline server/controllers/AuthController.js:108]: exclude this line from test

**Review by henryade — CHANGES_REQUESTED** [review]

File: server/routes/auth.js [hunk 13]
```diff
@@ -4,17 +4,23 @@ import AuthController from '../controllers/AuthController';
...
 // Sign up route
 auth.post(`${BASE_URL}/signup`, signUpValidation(), createUser);
-// Login route
+/* Login route */
 auth.post(`${BASE_URL}/login`, loginValidation(), login);
-// Verify User route
+/* Verify User route */
 auth.get(`${BASE_URL}/verify`, verifyToken, verifyUser);
+/* route responsible for sending password reset link to owners mail */ 
+auth.get(`${BASE_URL}/passwordreset`, EmailValidation(), resetpasswordEmail);
```
Comment (Lundii) [inline server/routes/auth.js]: Why are you using a get request here? Why is it not a post request.

because you are requesting for an access to change your password which would be sent to your mail,

Comment (Lundii) [reply server/routes/auth.js]: > Why are you using a get request here? Why is it not a post request.
> 
> because you are requesting for an access to change your password which would be sent to your mail,

Yes. But you added a body to the request. GET requests generally don't have a body with it. 

changed to post done

**Review by Ukhu — DISMISSED** [review]
The reset password feature should only be possible for users that have been verified, because, if they try testing the resetPasswordVerify route on Postman, which is gotten from the link sent to their email, it says they need to be verified to perform the operation. So I think the initial email that contains that link should not have been sent in the first place if they weren't verified.

resolved all users (verified or not verified will be able to reset their password)

File: server/routes/auth.js [hunk 14]
```diff
@@ -4,17 +4,23 @@ import AuthController from '../controllers/AuthController';
...
-// Login route
+/* Login route */
 auth.post(`${BASE_URL}/login`, loginValidation(), login);
-// Verify User route
+/* Verify User route */
 auth.get(`${BASE_URL}/verify`, verifyToken, verifyUser);
+/* route responsible for sending password reset link to owners mail */ 
+auth.post(`${BASE_URL}/passwordreset`, EmailValidation(), resetpasswordEmail);
+/* route responsible for password reset when mail has been sent */
+auth.patch(`${BASE_URL}/passwordresetVerify`, verifyToken, PasswordValidation(), resetPassword);
```
Comment (Lundii) [inline server/routes/auth.js:24]: Great job. So I tried testing it on m local before it. Here is what I did
* I signed up with my email
* I visited `POST api/v1/auth/passwordreset` with my mail
* I changed the the heroku url on the email sent to local host so I can test on my local
* an email was sent to me
* on clicking the link, I go a 404 not found error
* Tried to find out what the problem was, and I discovered this is a patch request. 
So my question
 * How is the user going to reset their password through the link if it is sending a patch request? keeping in mind that the link from their email will make a get request when they click it.

the link is not meant to be accessed by the user, it goes to a front end url which would append the appropriate http method and payload in the request body

File: server/helpers/capitalizeFirstLetter.js [hunk 15]
```diff
@@ -0,0 +1,8 @@
+const capitalizeFirstLetter = (word) => {
+  const formattedWord = word.toLowerCase().split(' ')
+    .map((f) => {
```
Comment (Nennyfills) [inline server/helpers/capitalizeFirstLetter.js:3]: why use f, consider using a meaningful word.

that's why i made the function capitalize first letter parameter a meaningful 'word '
the details of the function's implementation is abstracted away

**Review by Ukhu — CHANGES_REQUESTED** [review]
My observations after reviewing this PR;

1) The _securityDefinitions_ and _definitions_ properties in your Swagger Docs are duplicated.

2) I noticed you removed the documentation for the route where admin can activate and deactivate a user, and also the route where an admin can add a user.

Call to action:
Please kindly correct these discrepancies in your Swagger Docs.

done

File: server/helpers/messages.js [hunk 16]
```diff
@@ -15,19 +18,31 @@ const createUserMessage = (user, role, token) => {
...
     text: 'Hello Books',
     html: `  <div style="background-color:#7AA9EF; color:#fff; padding:20px"> <span style="text-align: left; margin-right: 10px">HB</span>
-             <span style= "text-align:center">Welcome to hello books ${user.firstName.toUpperCase()}</span></div>
-             <div style="margin-top:30px" >  You have been added as ${role === 'admin' ? 'an' : 'a'} ${role}. Your login details are email: ${user.email}
-             password: ${process.env.PASSWORD || 'setPassword'}.</div>
-             <div> Click on this link to confirm your email <a href=https://hello-books-staging.herokuapp.com/api/v1/auth/verify?token=${token}> LINK </a></div >
+             <span style= "text-align:center">Welcome to hello books ${capitalizeFirstLetter(user.firstName)}</span></div>
+             <div style="margin-top:30px" >You have been added as ${role === 'admin' ? 'an' : 'a'} ${role}. Your login details are email: ${user.email}
+             password: ${process.env.PASSWORD || 'setPassword'}</div>
+             <div> Click on this link to confirm your email <a href=${process.env.BASE_URL}/api/v1/auth/verify?token=${token}> LINK </a></div >
```
Comment (allebd) [inline server/helpers/messages.js:24]: Your URL is not in quotes as specified by Chris yesterday

**Review by Lundii — APPROVED** [review]
Good Job

**Review by Lundii — DISMISSED** [review]

**Review by Lundii — APPROVED** [review]

**Review by KvNGCzA — APPROVED** [review]
